spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/Mqttv5PahoMessageDrivenChannelAdapter.java

```diff
@@ -341,6 +380,46 @@ public void authPacketArrived(int reasonCode, MqttProperties properties) {
 
 	}
 
+	private void subscribeToAll() {
+		ApplicationEventPublisher applicationEventPublisher = getApplicationEventPublisher();
+		String[] topics = getTopic();
+		this.topicLock.lock();
+		try {
+			if (topics.length == 0) {
+				return;
+			}
+
+			int[] requestedQos = getQos();
+			if (this.mqttClient != null) {
+				this.mqttClient.subscribe(topics, requestedQos).waitForCompletion(getCompletionTimeout());
+			}
+			if (getClientManager() != null) {
+				MqttSubscription[] subscriptions = IntStream.range(0, topics.length)
+						.mapToObj(i -> new MqttSubscription(topics[i], requestedQos[i]))
+						.toArray(MqttSubscription[]::new);
+				MessageListener[] listeners = IntStream.range(0, topics.length)
+						.mapToObj(t -> new MessageListener())
+						.toArray(MessageListener[]::new);
+				getClientManager().getClient().subscribe(subscriptions, null, null, listeners, null)
```

Comment:
I wonder what is wrong with this method:
```
public IMqttToken subscribe(MqttSubscription[] subscriptions, IMqttMessageListener messageListener) throws MqttException;
```
and remember we talked about using a `MessageListener` even of an internally controlled client?

That's why I'm asking about that `this.mqttClient = getClientManager().getClient()`

Reply:
Initially I have used this method you mentioned - but it gives NPE. Yeah, I found it when ran my back-to-back integration tests with mosquitto. And it seems like the bug in the client lib.

> and remember we talked about using a MessageListener even of an internally controlled client?
Probably some other misunderstanding... sorry, I don't get it. `MessageListener` is the internal class of adapter which is a bridge: it implements the interface `IMqttMessageListener` of the library and reuses the existing logic of `messageArrived` method of the adapter class.

Reply:
The `messageArrived()` is not this class method. It is dictated by that `MqttCallback` we implement on this class.
With my finding it looks like this `messageArrived()` and `IMqttMessageListener` are exactly for the same purpose. 
With only a difference that `MqttCallback` is per client, but `IMqttMessageListener` is per request. Which is exactly what we need for our "shared client" use-case.
At the same time I want to see if the same `IMqttMessageListener` can be used for our case where we create client internally.

I also wonder if it really makes sense to `new MessageListener()` for every request, but not have it as a single one per channel adapter and just reuse for all the requests.
Kinda some minor, but still optimization.

Reply:
> org/eclipse/paho/mqttv5/client/MqttAsyncClient.java:1220 - passes an empty new MqttProperties()
and then org/eclipse/paho/mqttv5/client/MqttAsyncClient.java:1276 throws no such element

That is good catch! And I can confirm the bug is there.
Here is a bug report related: https://github.com/eclipse/paho.mqtt.java/issues/826